moped-editor/src/views/projects/projectView/ToolbarPlugin.js

```diff
@@ -0,0 +1,287 @@
+  CAN_REDO_COMMAND,
+  CAN_UNDO_COMMAND,
+  FORMAT_TEXT_COMMAND,
+  REDO_COMMAND,
+  SELECTION_CHANGE_COMMAND,
+  UNDO_COMMAND,
+  COMMAND_PRIORITY_LOW
+} from "lexical";
+import { mergeRegister } from "@lexical/utils";
+import { INSERT_ORDERED_LIST_COMMAND, INSERT_UNORDERED_LIST_COMMAND, REMOVE_LIST_COMMAND } from "@lexical/list";
+import { TOGGLE_LINK_COMMAND, $isLinkNode } from "@lexical/link";
+
+const richTextAction = {
+  bold: "bold",
+  italics: "italics",
+  underline: "underline",
+  number: "number",
+  bullet: "bullet",
+  link: "link",
+  divider: "divider",
+  undo: "undo",
+  redo: "redo",
+  clear: "clear"
+}
+
+const RICH_TEXT_OPTIONS = [
+  { id: richTextAction.bold, icon: <FormatBold />, label: "Bold" },
+  { id: richTextAction.italics, icon: <FormatItalic />, label: "Italics" },
+  { id: richTextAction.underline, icon: <FormatUnderlined />, label: "Underline" },
+  { id: richTextAction.divider },
+  {
+    id: richTextAction.number,
+    icon: <FormatListNumbered />,
+    label: "Numbered list",
+  },
+  {
+    id: richTextAction.bullet,
+    icon: <FormatListBulleted />,
+    label: "Bulleted list",
+  },
+  {
+    id: richTextAction.link,
+    icon: <Link />,
+    label: "Link",
+  },
+  { id: richTextAction.divider },
+  {
+    id: richTextAction.undo,
+    icon: <Undo />,
+    label: "Undo",
+  },
+  {
+    id: richTextAction.redo,
+    icon: <Redo />,
+    label: "Redo",
+  },
+  {
+    id: richTextAction.clear,
+    icon: <FormatClear />,
+    label: "Clear formatting",
+  },
+];
+
+const ToolbarPlugin = ({ noteAddSuccess, classes }) => {
+  const [editor] = useLexicalComposerContext();
+
+  const [disableMap, setDisableMap] = useState({
+    [richTextAction.undo]: true,
+    [richTextAction.redo]: true
+  });
+  const [selectionMap, setSelectionMap] = useState({});
+
+  // Iterate through selected node and parent nodes and returns any applicable list type
+  const checkListType = (selection, listType) => {
+    let hasListType = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((node) => {
+      const nodeParents = node.getParents();
+      nodeParents.forEach((parent) => {
+        if (parent.__listType === listType) {
+          hasListType = true;
+        }
+      })
+    })
+    return hasListType;
+  };
+
+  // Checks for selected list formatting, removes if already applied or adds if not applied
+  const handleListUpdate = useCallback((command, listType) => {
+    const isListApplied = selectionMap[listType];
+    isListApplied ?
+      editor.dispatchCommand(REMOVE_LIST_COMMAND) :
+      editor.dispatchCommand(command);
+  }, [editor, selectionMap]);
+
+  const checkLink = (selection) => {
+    let isLink = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      if ($isLinkNode(selectedNodeParent)) {
+        isLink = true
+      }
+    })
+    return isLink;
+  };
```

Comment:
i see some of these functions wrapped in useCallback and others not, so i'm curious what the distinction is. 

it looks like perhaps `checkLink` and  `checkListType` can be moved outside the component entirely to bring them out of the render cycle.

Reply:
good call on this! i moved these out of the component and that is definitely cleaner :)

Reply:
also to answer your first question i was receiving error messages on those particular functions stating they would continuously rerender so i had to wrap them in useCallbacks